includes/admin/extensions/abstract-extension.php

```diff
@@ -0,0 +1,209 @@
+	 *
+	 * @since 2.11.x
+	 * @var int
+	 */
+	protected $item_id;
+
+	/**
+	 * The required AA pass level.
+	 */
+	const PASS_LEVEL = Pass_Manager::PERSONAL_PASS_ID;
+
+	/**
+	 * The Extension Manager
+	 *
+	 * @var \EDD\Admin\Extensions\Extension_Manager
+	 */
+	protected $manager;
+
+	public function __construct() {
+		$this->manager = new \EDD\Admin\Extensions\Extension_Manager( static::PASS_LEVEL );
+	}
+
+	/**
+	 * Gets the configuration for the extension.
+	 *
+	 * @since 2.11.x
+	 * @param bool|int $item_id
+	 * @return array
+	 */
+	abstract protected function get_configuration( $item_id = false );
+
+	/**
+	 * Whether the extension is activated.
+	 *
+	 * @since 2.11.x
+	 * @return bool
+	 */
+	abstract protected function is_activated();
+
+	/**
+	 * Output the settings field (installation helper).
+	 *
+	 * @param array $args
+	 * @return void
+	 */
+	public function settings_field() {
+		if ( $this->is_activated() ) {
+			return;
+		}
+		$this->do_single_extension_card();
+	}
+
+	/**
+	 * Outputs a single extension card.
+	 *
+	 * @since 2.11.x
+	 * @param bool|int $item_id Optional: the individual extension product ID.
+	 * @return void
+	 */
+	public function do_single_extension_card( $item_id = false ) {
+		$config       = $this->get_configuration( $item_id );
+		$product_data = $this->get_product_data( $item_id );
+		if ( ! $product_data ) {
+			return;
+		}
+		$this->manager->do_extension_card(
+			$product_data,
+			$config,
+			$this->get_button_parameters( $config, $product_data ),
+			$this->get_link_parameters( $config, $product_data->info->title )
+		);
+	}
+
+	/**
+	 * Gets the product data for a specific extension.
+	 *
+	 * @param boolean $item_id
+	 * @return bool|object False if there is no data; API object if there is.
+	 */
+	public function get_product_data( $item_id = false ) {
+		$body = $this->get_api_body();
+		if ( empty( $body ) ) {
+			return false;
+		}
+		require_once EDD_PLUGIN_DIR . 'includes/admin/extensions/class-extensions-api.php';
+		$api          = new ExtensionsAPI();
+		$product_data = $api->get_product_data( $body, $this->item_id );
+		if ( ! $product_data ) {
+			return false;
+		}
+
+		return $this->item_id ? $product_data : $product_data[ $item_id ];
+	}
+
+	/**
+	 * Gets the array for the body of the API request.
+	 * Classes may need to override this (for example, to query a specific tag).
+	 *
+	 * @return array
+	 */
+	protected function get_api_body() {
+		return array(
+			'product' => $this->item_id,
+		);
+	}
+
+	/**
+	 * Gets the type for the button data-type attribute.
+	 * This is intended to sync with the Products API request.
+	 * Default is product.
+	 *
+	 * Really a shim for array_key_first.
+	 *
+	 * @param array $array
+	 * @return void
+	 */
+	private function get_type( array $array ) {
+		if ( function_exists( 'array_key_first' ) ) {
+			return array_key_first( $array );
+		}
+		foreach ( $array as $key => $unused ) {
+			return $key;
+		}
+
+		return 'product';
+	}
```

Comment:
I might have a concern about where this is getting the type information, but first, can you help me understand what the other options might be? I understand `$array` is coming from `get_api_body()`. What other options other than `product` would we have?

My potential concerns is that this method is requiring that the type---whatever that is---always be the first item in the `get_api_body()` array and I'm seeing that as a potential pitfall. I'd rather not have specific ordering requirements.

Reply:
So originally type was going to be `plugin` or `extension` and then as I started working with the products API, I felt like I could get more use out of it than just whether it was a plugin or something EDD specific--so the default type is `plugin` (outside of this class, I mean--but for managing the installation/activation process), and if it's anything else, the language at the end should change over to `extension`. But as I was working, I realized that I wanted to be able to set the product API result to an option, but I also want to delete it when the installation is done, because I don't want to leave an unnecessary option hanging in the database (that was one reason I liked a transient). So I'm using the first array item to determine the option name that I set when I query the products API, and then I delete that option on successful activation. So it could be `product_{$item_id}` or `tag_{$tag_id}` (that's what I have so far, as I was querying the entire email marketing tag, but only the recurring ID).

I don't think ordering should be an issue--the array is used this way twice (to extract the type and ID here and then to name the option when querying the API), but it's not filterable so the data should be the same for both places. But I'm open to suggestions too!